Optimizes the pipeline of basic mobs

LemonInTheDark opened the pull request:
## About The Pull Request

We use basic mobs a LOT, we've gotta make sure the system they use is as fast as it can be.
In service of this, I'm microing find_potential_targets, making incapicated event based so I can make able_to_run() event based, removing controller/process() in favor of timers and an idle handled alongside signals to react to movement from the pawn and its target.

### Potential behavior/capability changes
You can no longer modify a delay while waiting for it, at least not as easily
This might? make some of our movement logic slightly less responsive since it doesn't react to the world structure changing v well. I think? that's ok but idk I could see about catching that condition and feeding it into a process just for it

### Commits

For readability, the incap changes made this more noisy then I'd like so we're doin it like this
I'm leavin some out, I either think they're boring or I reverted them (the progressively increasing delay was one of these)

[Optimizes find_potential_targets some](https://github.com/tgstation/tgstation/commit/a8a09f6f1ace72c75c173804a4a9fdec71956b0f)

range() returns a LOT. it is real slow to filter through all of that manually.
Byond's type filtering is faster I hate to say it, so let's just use that.

[Sets up event based incapacitated](https://github.com/tgstation/tgstation/commit/c9bb106bba468789aab1232de2440fc2dac1db6c)

Rather then calling a proc every time we want to know if we're incap or
not, let's store all possible responses as a bitflag and update a var
whenever we change.

This way we can react to incap changes in an event based way!
Now to do the mass updating

[Implements var read for incapacitated for all consumers](https://github.com/tgstation/tgstation/pull/79498/commits/31894eb88023fa1a2c6749df070cbf252f317508)
[31894eb](https://github.com/tgstation/tgstation/pull/79498/commits/31894eb88023fa1a2c6749df070cbf252f317508)

Regexes used:
incapacitated() -> incapacitated
incapacitated(([^\)]*)) -> (incapacitated & $1)

I wrapped the flag bits with () cause I don't trust binary ops

[Implements event based able_to_run(), which should save a good portion off ai work](https://github.com/tgstation/tgstation/pull/79498/commits/c13d6320f233894870137cd7aa68ef45774bc807)
[c13d632](https://github.com/tgstation/tgstation/pull/79498/commits/c13d6320f233894870137cd7aa68ef45774bc807)

We've reached the point where we have a TON of basic mobs. we need to start microing the plumbing that they flow through so we can continue using them at scale

[Pulls out the bits of movable requiring behaviors that want to be run each tick into an event based system](https://github.com/tgstation/tgstation/pull/79498/commits/efb661f9e21c47e7fbf335036a2cbf0ebda69f9d)
[efb661f](https://github.com/tgstation/tgstation/pull/79498/commits/efb661f9e21c47e7fbf335036a2cbf0ebda69f9d)

Cheaper, mobs move way less then they process, plus this lets me get rid of the for loop here, which is gonna be real powerful just watch.
It's a bit fragile but workable imo

[Removes the behavior_cooldowns list, moves behavior timing to well, timers, clears a lot of shit out of behavior/process](https://github.com/tgstation/tgstation/pull/79498/commits/3d70a29ed0a9718609dfea95f0b4e078bd8b8641)
[3d70a29](https://github.com/tgstation/tgstation/pull/79498/commits/3d70a29ed0a9718609dfea95f0b4e078bd8b8641)

[fully yeets ai_behaviors as a processing subsystem, replaces it with ai_idle to handle idle processing, everything else is goin to fuckin timers bro](https://github.com/tgstation/tgstation/pull/79498/commits/e33794ac0eb451635092249320ebec176b347d7c)
[e33794a](https://github.com/tgstation/tgstation/pull/79498/commits/e33794ac0eb451635092249320ebec176b347d7c)

LemonInTheDark committed:
Optimizes find_potential_targets some

range() returns a LOT. it is real slow to filter through all of that
manually.
Byond's type filtering is faster I hate to say it, so let's just use
that.

LemonInTheDark committed:
Sets up event based incapacitated

Rather then calling a proc every time we want to know if we're incap or
not, let's store all possible responses as a bitflag and update a var
whenever we change.

This way we can react to incap changes in an event based way!
Now to do the mass updating

LemonInTheDark committed:
Implements var read for incapacitated for all consumers

Regexes used:
incapacitated\(\) -> incapacitated
incapacitated\(([^\)]*)\) -> (incapacitated & $1)

I wrapped the flag bits with () cause I don't trust binary ops

LemonInTheDark committed:
adds linter hint

LemonInTheDark committed:
updates all the bits the regex fucked

LemonInTheDark committed:
doc comment

LemonInTheDark committed:
Implements event based able_to_run(), which should save a good portion off ai work

We've reached the point where we have a TON of basic mobs. we need to
start microing the plumbing that they flow through so we can continue
using them at scale

LemonInTheDark committed:
micros /datum/ai_controller/process

LemonInTheDark committed:
Implements a progressively increasing delay between targetting attempts

This will very likely end up running on a loop forever. We should do
what we can to ensure it doesn't chew cpu on nothing.

Frankly it might be a better pattern to keep track of things moving in
its rough area if it gets stuck in this loop for too long, and then just
let event based do its job, but idk this is easier. Less sure of it tho,
this is the change I'm least confident in. can walk back if you'd like

LemonInTheDark committed:
Pulls out the bits of movable requiring behaviors that want to be run each tick into an event based system

Cheaper, mobs move way less then they process, plus this lets me get rid
of the for loop here, which is gonna be real powerful just watch. It's a
bit fragile but workable imo

LemonInTheDark committed:
removes silly

LemonInTheDark committed:
Removes the behavior_cooldowns list, moves behavior timing to well, timers, clears a lot of shit out of behavior/process

LemonInTheDark committed:
removes the able_to_run check from process, unneeded

LemonInTheDark committed:
fully yeets ai_behaviors as a processing subsystem, replaces it with ai_idle to handle idle processing, everything else is goin to fuckin timers bro

LemonInTheDark committed:
missed a few spots

LemonInTheDark commented:
If the cl bot wipes my pr desc I will cry
profiles tomorrow

LemonInTheDark commented:
I am going to go cry

LemonInTheDark committed:
fucking git

Kapu1178 commented on code/modules/mob/living/living.dm:
```diff
@@ -457,31 +457,23 @@
-/mob/living/incapacitated(flags)
-	if(HAS_TRAIT(src, TRAIT_INCAPACITATED))
-		return TRUE
+/mob/living/setup_incapacitated()
+	RegisterSignals(src, list(SIGNAL_ADDTRAIT(TRAIT_INCAPACITATED), SIGNAL_REMOVETRAIT(TRAIT_INCAPACITATED)), PROC_REF(update_incapacitated))
+	RegisterSignals(src, list(SIGNAL_ADDTRAIT(TRAIT_RESTRAINED), SIGNAL_REMOVETRAIT(TRAIT_RESTRAINED)), PROC_REF(update_incapacitated))
+	RegisterSignals(src, list(SIGNAL_ADDTRAIT(TRAIT_STASIS), SIGNAL_REMOVETRAIT(TRAIT_STASIS)), PROC_REF(update_incapacitated))
```
code\modules\mob\living\init_signals.dm

LemonInTheDark committed:
ensures behaviors will requeue their timer if we don't finish them

LemonInTheDark committed:
Refactors incapacitated to support multiple flags correctly

Includes a macro to make it easier to reason about.
Also moved all my signal stuff to init_signals, rip
setup_incapacitated()

LemonInTheDark committed:
fixes up some bugs, improves the signals override error message

LemonInTheDark committed:
Implements a get_cooldown() proc to get around dumb manual overrides, and empower me to optimize the findtarget logic

LemonInTheDark committed:
autodoc updates for timers

LemonInTheDark committed:
Properly implements ai behavior handling

Behaviors are handled sequentually, not all at once

LemonInTheDark committed:
Makes planned behaviors a flat list, fixes a bug w removal that dropped args

LemonInTheDark committed:
Implements a signal sent when we repick behaviors

LemonInTheDark committed:
updates the previous to actually use planned behaviors

LemonInTheDark committed:
Refactors fields to pass along the old turf in entered and new in exited

LemonInTheDark committed:
Implements a proximity monitor that will notify an ai controller if the pawn moves or if somethings moves in range

LemonInTheDark committed:
Implements clearing support for lists, this seems important

LemonInTheDark committed:
Adds a signal to behavior queuing that sends the args along

LemonInTheDark committed:
code should compile

LemonInTheDark committed:
more error fixing

LemonInTheDark committed:
error fixes again

LemonInTheDark committed:
Implements timer based ai properly for once

Rather then doing actions sequentially, blocking one until the one
before is done, we do them as requested, and just ensure that we never
run more then once at once

LemonInTheDark committed:
prevents doing work post proximity monitor qdel, there are other ways of doing this but I think? this is faster?

LemonInTheDark committed:
de lazylists ai controllers. The motivation was good but it doesn't actuallly come into play ever

LemonInTheDark committed:
adds bespoke logic for handling init on proximity monitors, to ensure it works with entered

LemonInTheDark committed:
Implements a special kind of field to handle ai targeting

If we run targeting and don't like, find anything, we should setup a
field that listens for things coming near us and then handle those
things as we find them.

This incurs a slight startup cost but saves so much time on the churn
of constant costs

Note:
We should also work to figure out a way to avoid waking ais if none is
near them/they aren't doing anything interesting

We don't need to do that immediately this acts as somewhat of a stopgap
(and would be good regardless) but it is worth keeping in mind)

LemonInTheDark committed:
Adds return flags to perform, properly implements behavior queuing

Behaviors need to queue one at a time, with behaviors higher up the list
taking priority if cooldowns are shared

This code implements that

As an optimization, it is useful to know for sure if finish_action has
been called by the end of a perform() call. Also, people were just
forgetting to feed it args.

In service of those two issues, let's add return flags that call
finish_action.

In addition, perform() controls when delay is applied to actions and
when it isn't.

So let's use perform()'s return flags to signal that, rather then
faffing about with doing it inside the proc like we did before.

As an optimization over the timer system, if we have a behavior that
wants to run on the next process, then we should well, process() rather
then using a timer (under the assumption it'll keep wanting that)

LemonInTheDark committed:
Optimizes connect_range and promxity_monitors

We know what turfs exist before and after a move
We can use this information to prevent trying to update turfs we don't
care about.

This is important because post these changes mobs with fields will be
moving a lot more, so it's gotta be cheap

LemonInTheDark committed:
Adds modify_cooldown, uses it to optimize find_potential_targets further

No sense running the behavior if we're just waiting on its output, so
let's run it once a minute just in case, then push an update instantly
if we find something

LemonInTheDark commented:
Much better/faster, just not happy with behavior quite yet

Iamgoofball commented:
stale b gone

LemonInTheDark committed:
Fixes up some code/broken behavior

LemonInTheDark committed:
/datum/ai_behavior/find_potential_targets/perform(seconds_per_tick, datum/ai_controller/controller, target_key, targetting_datum_key, hiding_location_key)
Merge branch 'master' of https://github.com/tgstation/tgstation into ai_ops

LemonInTheDark committed:
Cleans up existing code to support my changes

Replaces outdated names

Adds a setter for bot mode flags so I can react to it changing and set
able_to_run in reaction

Updates all uses of the incapacitated proc to the macro

Updates all uses of finish_action to the return pattern

Adds a qdeleted check to jps movement, if it fails we might have just
qdeleted after all

LemonInTheDark commented:
Popping this up as a TM to make sure behavior is roughly right (since I'm changing a bunch of shit)

LemonInTheDark committed:
bad space

LemonInTheDark committed:
Fixes unit tests, adds timer getter tenatively

LemonInTheDark committed:
whoop

LemonInTheDark committed:
two warns bite the dust

MrMelbert commented:
`incapacitated` does not seem to remotely update accurately

Putting someone on stasis does not update the var, grabbing some aggressively does not update the var, but if you aggro-grab someone in crit and heal them out of crit they will be stuck incapacitated forever

![image](https://github.com/tgstation/tgstation/assets/51863163/50d81bc7-1612-46b4-9923-efc059cd96bf)

spookydonut commented on code/controllers/subsystem/timer.dm:
```diff
@@ -681,11 +682,36 @@ SUBSYSTEM_DEF(timer)
+ * * timer_subsystem - Optional, the subsystem to pull from. Replaced w SStimer if none is given
+ */
+/proc/gettimer(id, datum/controller/subsystem/timer/timer_subsystem)
+	if (!id)
```
maybe add like #IFNDEF DEBUG to_chat(world, "someone being a dickhead with gettimer()(")

LemonInTheDark replied:
s mostly just a copypasta from the other procs, at this point I'd stacktrace tbh

spookydonut commented on code/datums/ai/basic_mobs/basic_ai_behaviors/targeting.dm:
```diff
@@ -4,11 +4,15 @@
+
+/datum/ai_behavior/find_potential_targets/get_cooldown(datum/ai_controller/cooldown_for)
+	if(cooldown_for.blackboard[BB_FIND_TARGETS_FIELD(type)])
+		return 60 SECONDS
```
feels arbitary?  players might think its broken with such a long delay

LemonInTheDark replied:
In theory the delay doesn't really matter since any new mobs will trigger the field and get detected, I think? this was just a backup, tho I could be misremembering the logic.

LemonInTheDark commented:
guh
